### DISCUSSION TITLE
Initial data sync workflow and UX redesign

### DISCUSSION THREAD

**Discussion by ndegwamartin** [opensrp/fhircore#2678] [Ideas]
**Overview**
Currently the implementation of initial sync is such that once it fails to complete it  blocks the user with the sync progress dialog even after all the default retries are complete.

From [the conversation here](https://github.com/opensrp/fhircore/issues/2575) sounds like the consensus is to block the user from using the app until all data in the initial sync has successfully been synced down.

The above leads to a poor user experience given that the app appears stuck with no possible user actions. The idea is to come up with the final design and subsequent implementation to improve the UX above.

Here is a collation of discussion points ideas that have gone into this conversation.
 
[@joyce-x-chen ]
- If initial sync fails, how many times does it retry before it stops?
- When initial sync restarts, does it resume from where it left off or start again from scratch?
Desired UX cc **design-team @pms, @tpms**
- User is informed that sync has started
  - While syncing, user is aware of the progress of sync (i.e. sees the sync % complete)
- If sync fails, user is 
  - Informed that sync is incomplete,
  - Given a reason why and/or instructions on what to do next, and
  - Informed that the app cannot be used until initial sync is completed
- User is able to reinitiate the initial sync process manually
- [If applicable] User is warned of what not to do
Can we try to think of reasons why initial sync would fail so we can provide more context/instructions? For example:
- Server unavailable. Try again later
- Connectivity issue
  - No internet connection. Check your connection and try again
  - Unstable internet connection. Check your connection and try again.

When is the next retry cycle ? Is it the same as normal remote sync, i.e. every 15 minutes and what is the “start” of the 15 minute interval? The generic sync failed error looks ok for this until we have more time to workshop and test other error messages.

- For initial sync, it retries 3 times then waits till the next retry cycle. It then retries from where it left off. When there’s an error to do with authentication we show the Snackbar message Sync failed because the authentication credentials are invalid. For any issue other than this we show Sync failed. Check internet connection or try again later.

**[@ndegwamartin ]** 
It should be the same periodic sync triggering process thus recurring every 15mins after a user logs in and enters the pin

**[@ellykits ]**
That's correct periodic sync is by default scheduled to run every 15 mins; the duration can however be overridden via app configuration but should not be less than 15mins as required by Android system.

**[@joyce-x-chen ]**
What are the user’s possible actions after seeing the sync failed error?
- Can they close the error? If yes, what happens then?
- Are they forced to retry resync?
- What happens if they leave the app and come back?
I have some apps on my phone that force an update before I can use it…and it usually stays on a “startup” page before allowing me to enter the app. I wonder if it’d be more intuitive for initial sync to happen during the initial login process rather than AFTER successfully logging in. So the user stays on the login page during initial sync, and only after successful initial sync, can they enter the app. Cc: @pld 

**[@ndegwamartin ]**
For the items above
- Can they close the error? If yes, what happens then? - No, the sync progress dialog blocks all further interactions with the user (i.e. after we merge this PR [#2608](https://github.com/opensrp/fhircore/pull/2608) - see issue on current implementation here [2606](https://github.com/opensrp/fhircore/issues/2606))
- Are they forced to retry re-sync? - No - it kicks in automatically after 15 mins
- What happens if they leave the app and come back? - Picks up where it left off (minor bug though reported by Anne where it resets to 0 if there’s no connection)

On the suggestion for sync before the Login page redirects I think I agree it can lead to better UX. The user will never get past the Log in page if sync doesn’t complete so it won’t be confusing for the user on whether they should interact with the app or not.

**[@dubdabasoduba ]**
I agree on keeping the user on the login page until the initial sync is complete.  We however need to account for the following things.
- Giving the user a visual indication of what is happening. This visual indication should have the following items.
   - Communicate when login is complete and sync has started. This tells the users what is actually going on.
   - Communicate the sync percentages.
   - Communicate any errors

**[@AnnieMungai ]**
@ndegwamartin if sync fails - can we add the different reasons for the failure? Example
Sync failed- no internet connection.
Sync failed - XYZ other reason. Can we know the reasons for sync failure and give brief instructions on what the user should do?

**[@samkanga ]**
I am in support of blocking the app use until we get a successful sync retry or manual sync. We are bound to run into more issues if we allow user to proceed without a successful initial sync or sync retry. These would include:
- Chance of creating duplicates in cases where a patient details cannot be found.
- Missing tasks where generated tasks are part of the details not synced down.

**[@dubdabasoduba ]**
I agree on keeping the user on the login page until the initial sync is complete.  We however need to account for the following things.
- Giving the user a visual indication of what is happening. This visual indication should have the following items.
  - Communicate when login is complete and sync has started. This tells the users what is actually going on.
  - Communicate the sync percentages.
  - Communicate any errors

**Resources**
- https://github.com/opensrp/fhircore/issues/2575
- https://github.com/opensrp/fhircore/issues/2606

**Comment by ndegwamartin** [comment]
> On the suggestion for sync before the Login page redirects I think I agree it can lead to better UX. The user will never get past the Log in page if sync doesn’t complete so it won’t be confusing for the user on whether they should interact with the app or not.

In reference to my comment above, I think we can have the new workflow take the said approach. That is, on app install and launch:
1. Show the app settings/application id page
2. Enter application I'd and navigate to login page
3. Login in and start syncing down. Render an Android native progress dialog to show progress
4. If sync down fully succeeds navigate to the Register page. 
5. If sync fails remain on the Login Page with the appropriate error message _(and maybe next user actions)_

**Attached is an example of how this progress dialog would look like:**
<img width="684" alt="Screenshot 2023-08-14 at 14 37 11" src="https://github.com/opensrp/fhircore/assets/10017086/42b26410-53d5-4f07-9711-bfa1711ee917">

<img width="714" alt="Screenshot 2023-08-14 at 14 36 39" src="https://github.com/opensrp/fhircore/assets/10017086/0f5fbfcc-8ebe-467f-8cbb-a04da81f3e6f">


**Here are some other great ideas.**
<img width="343" alt="Screenshot 2023-08-14 at 14 18 13" src="https://github.com/opensrp/fhircore/assets/10017086/42a92d2a-91a5-4c23-805e-3c6d12c21b27">
<img width="730" alt="Screenshot 2023-08-14 at 14 18 39" src="https://github.com/opensrp/fhircore/assets/10017086/0543e20b-8a85-49a9-ba1d-d5eb251b15b9">
<img width="535" alt="Screenshot 2023-08-14 at 14 18 50" src="https://github.com/opensrp/fhircore/assets/10017086/510e4195-7269-48a7-9916-45164e1b4abb">

cc @HenryRae @pld @ellykits

**Comment by HenryRae** [comment]
> [@dubdabasoduba ]
I agree on keeping the user on the login page until the initial sync is complete. We however need to account for the following things.

User shouldn't at any point be kept on login screen until sync completes. We can log the user in, show some of the synced data and progress of the remaining data to be synced. We should avoid to keep the user waiting, it's a very frustrating User experience. 

> [@samkanga ]
I am in support of blocking the app use until we get a successful sync retry or manual sync. We are bound to run into more issues if we allow user to proceed without a successful initial sync or sync retry. 

We should never block user from accessing the app because of a sync. We can should some already synced data as the rest keeps syncing for better UX. A lot of users might get frustrated with this kind of experience especially if they have very slow internet connection.

@ndegwamartin what timelines do you have on this, I can take a look at it in detail and share more thoughts when I get a moment.

**Reply by ndegwamartin** [reply]
@HenryRae we don't have timelines defined yet but we may need to come up with some cc @dubdabasoduba @pld 

For the blocking UX the approach is in line with the attached ticket here [#2602](https://github.com/opensrp/fhircore/issues/2606), please review. 

But to summarise the idea is to block the user from interaction unless everything is completed. The fact that this happens one time only during initial sync might alleviate the bad UX. The other option is to indeed let the user proceed to register page after login but block them from doing anything until it completes which is what we actually do now.

**Reply by pld** [reply]
How do we indicate it an initial sync is still in-progress when they're on the action-blocked register page?

**Reply by joyce-x-chen** [reply]
@HenryRae Just want to clarify that the use case here is not routine sync of data between the device and the server. It's only for the initial sync, when the app fetches all of the relevant configs and downloads the data that the user should have access to. The reason there is discussion about blocking use of the app until all sync is complete is because, if they are able to use the app before all files and data have been successfully downsynced, they will experience missing data and missing tasks and inaccurate calculations. That has the potential to decrease their confidence in the app's integrity, which could be more damaging than the frustration that they encounter once and only during the initial login and sync. 

In many program contexts, the devices are set up by someone other than the frontline health worker. So the frustration of being stuck at the login page is likely to be borne by a program staff member that will have a better understanding of what is happening and why, and it will also ensure that they continue with device setup until they successfully pass the login page. 

If they are allowed to progress to the register page, it is easy to accidentally mistaken/assume that sync was successfully completed. And by the time the device makes it to the field, there may not be reliable internet access to complete initial sync.

Sorry for the long-winded elaboration, but wanted to clear this up in case your reservations were due to a misunderstanding.

**Reply by pld** [reply]
Is the sync split between the app configs and the data? If not we should be able to do that. Then we can bring the user in but have all the registers be non-interactive and on the registers show a shimmer with the loading.

Another issue with staying on the log-in page is it's unclear to the user if the login succeeded, like that your auth details are correct. Also not sure how those login details are saved if we stay on the login page, like if you force quit during initial sync, then reopen app, does the initial sync start again with the saved login details but on the login page? I guess that's fine, but maybe a little confusing to the user.

**Reply by ndegwamartin** [reply]
@pld for your first question checkout a screenshot on [the link here](https://github.com/opensrp/fhircore/pull/2152)

**Reply by pld** [reply]
Cool, I think as is in that screenshot is better than sitting on the login page. Even better would be to more selectively disable stuff in stead of the whole app, but that gets very tricky, and maybe there's not enough you could do during an initial sync to make that worthwhile

**Reply by HenryRae** [reply]
:smiling_face_with_tear: Are we able to allow user maybe access features that have been synced as others are still in sync, related to @pld's question of split sync? 
Are there potential ways we could tone down on the sync time(would be a major issue when connection is slow)?

**Reply by ndegwamartin** [reply]
I concur with _pld_, that would _(most probably)_ be complex to implement, maintain and scale

**Reply by pld** [reply]
yes, we're actively looking into reducing the sync time

**Comment by ndegwamartin** [comment]
Here are some related guidelines by OHS that can inform the discussion here as well 
- https://developers.google.com/open-health-stack/design/offline-sync-guideline
